### PR TITLE
[JUJU-4170] Create mux terraform framework

### PR DESCRIPTION
**Pull request by hmlanigan**
## Description

The first step in migrating the juju terraform provider from the terraform plugin sdk2 to the terraform plugin framework. Created a framework style provider to sit beside the current sdk2 style provider, containing the same functionality. The 2 providers have been added to a MUX available for transitions between types of provider styles until all of the resources and data sources can also be transformed. Then the MUX will be removed.

To enable the MUX, the sdk provider's ConfigureContextFunc was updated to get default values for the provider config as DefaultFunc is no longer part of the schema. The method to find default provider config was updated to do it all at once, rather than individually. As well, it is shared with the new framework provider.

## Type of change

Maintenance.

## Environment

- Juju controller version: 2.9.44
- Terraform version: v1.5.2

## QA steps

No change to current behavior. All acceptance tests should run.

# Additional notes

Additional PRs transforming the current resources and data sources are on the way.

More tests will be added as work continues. The framework tests are significantly different from the current ones.

### CODE REVIEW

**Comment by jnsgruk** [comment]
Nice, thanks @hmlanigan!

**Review by cderici — APPROVED** [review]
Some small tweaks as commented, o/w LGTM 👍 (also needs a `go generate ./...`)

Some of the tests have failed running locally, so expect some failures (e.g. `TestProviderConfigureUsernameFromEnv`).

File: internal/provider/provider.go [hunk 1]
```diff
@@ -134,16 +161,229 @@ func configure(version string, p *schema.Provider) func(context.Context, *schema
...
+func (p *jujuProvider) Metadata(_ context.Context, _ frameworkprovider.MetadataRequest, resp *frameworkprovider.MetadataResponse) {
+	resp.TypeName = "juju"
+	resp.Version = p.version
+}
+
+// Schema returns the schema for this provider, specifically
+// it defines the juju controller config necessary to create
+// a juju client.
+func (p *jujuProvider) Schema(_ context.Context, _ frameworkprovider.SchemaRequest, resp *frameworkprovider.SchemaResponse) {
+	resp.Schema = frameworkschema.Schema{
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply internal/provider/provider.go:243]: The schema types between the sdk and framework are different and not compatible.

File: main.go [hunk 2]
```diff
@@ -28,20 +35,37 @@ var (
...
+		serveOpts = append(serveOpts, tf5server.WithManagedDebug())
 		zerolog.SetGlobalLevel(zerolog.DebugLevel)
 	}
 
-	opts := &plugin.ServeOpts{
-		ProviderFunc: provider.New(version),
-		ProviderAddr: "registry.terraform.io/canonical/juju",
-		Debug:        debugMode,
+	if err = tf5server.Serve(
+		"registry.terraform.io/<namespace>/<provider_name>",
```
[the remark this answers is not in the document]
Comment (hmlanigan) [reply main.go:65]: fixed.

**Pull request merged**